TypeRelating emit WellFormed, not generalize

lcnr opened the pull request:
fixes https://github.com/rust-lang/trait-system-refactor-initiative/issues/250

r? BoxyUwU

rust-timer commented:
Finished benchmarking commit (20a081aa36c390f34fa701ef245698110b60e282): [comparison URL](https://perf.rust-lang.org/compare.html?start=9312cd6d38396a13a08be73bb589fb1cf3dd32e6&end=20a081aa36c390f34fa701ef245698110b60e282&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please do so in sufficient writing along with `@rustbot label: +perf-regression-triaged`. If not, please fix the regressions and do another perf run. If its results are neutral or positive, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=9312cd6d38396a13a08be73bb589fb1cf3dd32e6&end=20a081aa36c390f34fa701ef245698110b60e282&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.3%  | [2.0%, 2.8%]   | 6     |
| Regressions ❌ <br /> (secondary)  | 5.7%  | [0.1%, 15.2%]  | 6     |
| Improvements ✅ <br /> (primary)   | -1.3% | [-1.3%, -1.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 1.8%  | [-1.3%, 2.8%]  | 7     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=9312cd6d38396a13a08be73bb589fb1cf3dd32e6&end=20a081aa36c390f34fa701ef245698110b60e282&stat=max-rss)
<details>
<summary>Results (secondary 1.2%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.9%  | [3.9%, 3.9%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.4% | [-1.4%, -1.4%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=9312cd6d38396a13a08be73bb589fb1cf3dd32e6&end=20a081aa36c390f34fa701ef245698110b60e282&stat=cycles:u)
<details>
<summary>Results (primary 2.3%, secondary 10.1%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range         | count |
|:----------------------------------:|:-----:|:-------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.3%  | [2.2%, 2.5%]  | 3     |
| Regressions ❌ <br /> (secondary)  | 10.1% | [2.6%, 14.9%] | 3     |
| Improvements ✅ <br /> (primary)   | -     | -             | 0     |
| Improvements ✅ <br /> (secondary) | -     | -             | 0     |
| All ❌✅ (primary)                 | 2.3%  | [2.2%, 2.5%]  | 3     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=9312cd6d38396a13a08be73bb589fb1cf3dd32e6&end=20a081aa36c390f34fa701ef245698110b60e282&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 477.625s -> 479.621s (0.42%)
**Artifact size**: 391.41 MiB -> 391.37 MiB (-0.01%)

lcnr committed:
emit WF goals in relate, not in generalize

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

lcnr committed:
MIR typeck: invariant ctxt fast path

This removes variance information from some diagnostics. However,
that variance information is not actually relevant here. Casting
`*const dyn Cat<'a>` to `*const S<dyn Cat<'static>>` is an error
regardless of whether `S` requires its argument to be invariant.
Wide-pointer casts always require the trait object arguments to be
invariant.

lcnr commented:
@bors try @rust-timer queue

rust-timer commented:
Queued 898f13cd81a0ddff99ba0cc248c43c60d4738716 with parent d682af88a57b0045f8348507682c16c6160b522d, future [comparison URL](https://perf.rust-lang.org/compare.html?start=d682af88a57b0045f8348507682c16c6160b522d&end=898f13cd81a0ddff99ba0cc248c43c60d4738716).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.5 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (898f13cd81a0ddff99ba0cc248c43c60d4738716): [comparison URL](https://perf.rust-lang.org/compare.html?start=d682af88a57b0045f8348507682c16c6160b522d&end=898f13cd81a0ddff99ba0cc248c43c60d4738716&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=d682af88a57b0045f8348507682c16c6160b522d&end=898f13cd81a0ddff99ba0cc248c43c60d4738716&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.2%  | [0.2%, 0.2%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-1.3%, -0.1%] | 3     |
| Improvements ✅ <br /> (secondary) | -0.8% | [-1.4%, -0.2%] | 12    |
| All ❌✅ (primary)                 | -0.5% | [-1.3%, -0.1%] | 3     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d682af88a57b0045f8348507682c16c6160b522d&end=898f13cd81a0ddff99ba0cc248c43c60d4738716&stat=max-rss)
<details>
<summary>Results (primary 1.9%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.6%  | [4.6%, 4.6%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.8% | [-0.8%, -0.8%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 1.9%  | [-0.8%, 4.6%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d682af88a57b0045f8348507682c16c6160b522d&end=898f13cd81a0ddff99ba0cc248c43c60d4738716&stat=cycles:u)
<details>
<summary>Results (secondary 3.7%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.7% | [2.1%, 4.8%] | 4     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=d682af88a57b0045f8348507682c16c6160b522d&end=898f13cd81a0ddff99ba0cc248c43c60d4738716&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 472.838s -> 476.429s (0.76%)
**Artifact size**: 388.40 MiB -> 388.47 MiB (0.02%)

lcnr commented on tests/ui/cast/ptr-to-trait-obj-different-regions-id-trait.current.stderr:
```diff
@@ -6,10 +6,6 @@ LL | fn m<'a>() {
-   |
-   = note: requirement occurs because of the type `S<dyn Cat<'_>>`, which makes the generic argument `dyn Cat<'_>` invariant
-   = note: the struct `S<T>` is invariant over the parameter `T`
-   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
```
Looked into this and think that removing that diagnostics information is actually good. The fact that the generic parameter of `S` is invariant doesn't actually matter here. The cast itself already requires equality of the trait object.

BoxyUwU commented on compiler/rustc_infer/src/infer/relate/generalize.rs:
```diff
@@ -391,10 +376,8 @@ impl<'tcx> Generalizer<'_, 'tcx> {
-    /// generalizing an alias. This has to set `has_unconstrained_ty_var`
-    /// if we're currently in a bivariant context.
+    /// generalizing an alias.
     fn next_ty_var_for_alias(&mut self) -> Ty<'tcx> {
```
take a non mut ref?

BoxyUwU reviewed (approved):
r=me with or without nit

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/148823/6840234806e4a57e14112a1137783c885b7a23db..8d77bbe94431b335e8d1f58f5c1bfeeb09718d8c)*

BoxyUwU commented:
@rustbot author

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/rust/actions/runs/19966791946/job/57268510134)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust/actions/runs/19966791946/job/57268510134","builder_name":"checks-actions"} -->

rust-timer commented:
Finished benchmarking commit (b4f1098e10e71139bbdc08df5ea8183e68fc0955): [comparison URL](https://perf.rust-lang.org/compare.html?start=36b2369c91d32c2659887ed6fe3d570640f44fd2&end=b4f1098e10e71139bbdc08df5ea8183e68fc0955&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Our benchmarks found a performance regression caused by this PR.
This might be an actual regression, but it can also be just noise.

**Next Steps**:

- If the regression was expected or you think it can be justified,
please write a comment with sufficient written justification, and add
`@rustbot label: +perf-regression-triaged` to it, to mark the regression as triaged.
- If you think that you know of a way to resolve the regression, try to create
a new PR with a fix for the regression.
- If you do not understand the regression or you think that it is just noise,
you can ask the `@rust-lang/wg-compiler-performance` working group for help (members of this group
were already notified of this PR).

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=36b2369c91d32c2659887ed6fe3d570640f44fd2&end=b4f1098e10e71139bbdc08df5ea8183e68fc0955&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.2%  | [0.0%, 0.5%]   | 14    |
| Improvements ✅ <br /> (primary)   | -1.3% | [-1.3%, -1.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -0.2% | [-0.3%, -0.1%] | 8     |
| All ❌✅ (primary)                 | -1.3% | [-1.3%, -1.3%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=36b2369c91d32c2659887ed6fe3d570640f44fd2&end=b4f1098e10e71139bbdc08df5ea8183e68fc0955&stat=max-rss)
<details>
<summary>Results (primary 1.0%, secondary 1.9%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.0%  | [1.0%, 1.0%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 4.0%  | [1.6%, 6.6%]   | 9     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.3% | [-2.4%, -0.8%] | 6     |
| All ❌✅ (primary)                 | 1.0%  | [1.0%, 1.0%]   | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=36b2369c91d32c2659887ed6fe3d570640f44fd2&end=b4f1098e10e71139bbdc08df5ea8183e68fc0955&stat=cycles:u)
<details>
<summary>Results (primary 3.0%, secondary -1.2%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.0%  | [2.8%, 3.3%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 4.0%  | [2.6%, 6.1%]   | 3     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.8% | [-6.6%, -1.8%] | 6     |
| All ❌✅ (primary)                 | 3.0%  | [2.8%, 3.3%]   | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=36b2369c91d32c2659887ed6fe3d570640f44fd2&end=b4f1098e10e71139bbdc08df5ea8183e68fc0955&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.0%, secondary -0.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.0% | [-0.0%, -0.0%] | 1     |
| Improvements ✅ <br /> (secondary) | -0.0% | [-0.0%, -0.0%] | 1     |
| All ❌✅ (primary)                 | -0.0% | [-0.0%, -0.0%] | 1     |
</details>

**Bootstrap**: 468.958s -> 471.737s (0.59%)
**Artifact size**: 386.84 MiB -> 388.97 MiB (0.55%)
